File: my-solutions/0394-decode-string/solution.cpp
class Solution {
public:
    string decodeString(string s) {
        vector<int> counts;
        vector<string> frags;
        string curr;
        int num = 0;

        for (char c : s) {
            if (isdigit(c)) {
                num = num * 10 + (c - '0');
            } else if (c == '[') {
                counts.push_back(num);
                frags.push_back(curr);
                num = 0;
                curr.clear();
            } else if (c == ']') {
                int k = counts.back();
                counts.pop_back();
                string temp;
                temp.reserve(curr.size() * k);
                while (k--) {
                    temp += curr;
                }
                curr = frags.back() + temp;
                frags.pop_back();
            } else {
                curr.push_back(c);
            }
        }

        return curr;
    }
};

File: my-solutions/0394-decode-string/solution.py
class Solution:
    def decodeString(self, s: str) -> str:
        scalars = []
        strs = []
        curr_str = ""
        curr_num = 0

        for c in s:
            if c.isdigit():
                curr_num = curr_num * 10 + int(c)
            elif c == '[':
                scalars.append(curr_num)
                strs.append(curr_str)
                curr_num = 0
                curr_str = ""
            elif c == ']':
                repeat = scalars.pop()
                prev_str = strs.pop()
                curr_str = prev_str + curr_str * repeat
            else:
                curr_str += c
        
        return curr_str
        
